kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala

```diff
@@ -50,88 +50,133 @@ class FlinkProcessBuilder(
 
   val flinkHome: String = getEngineHome(shortName)
 
+  val flinkExecutable: String = {
+    Paths.get(flinkHome, "bin", FLINK_EXEC_FILE).toFile.getCanonicalPath
+  }
+
   override protected def module: String = "kyuubi-flink-sql-engine"
 
   override protected def mainClass: String = "org.apache.kyuubi.engine.flink.FlinkSQLEngine"
 
   override def env: Map[String, String] = conf.getEnvs +
     (FLINK_PROXY_USER_KEY -> proxyUser)
 
+  override def clusterManager(): Option[String] = Some("yarn")
+
   override protected val commands: Array[String] = {
     KyuubiApplicationManager.tagApplication(engineRefId, shortName, clusterManager(), conf)
-    val buffer = new ArrayBuffer[String]()
-    buffer += executable
-
-    val memory = conf.get(ENGINE_FLINK_MEMORY)
-    buffer += s"-Xmx$memory"
-    val javaOptions = conf.get(ENGINE_FLINK_JAVA_OPTIONS)
-    if (javaOptions.isDefined) {
-      buffer += javaOptions.get
-    }
 
-    buffer += "-cp"
-    val classpathEntries = new java.util.LinkedHashSet[String]
-    // flink engine runtime jar
-    mainResource.foreach(classpathEntries.add)
-    // flink sql client jar
-    val flinkSqlClientPath = Paths.get(flinkHome)
-      .resolve("opt")
-      .toFile
-      .listFiles(new FilenameFilter {
-        override def accept(dir: File, name: String): Boolean = {
-          name.toLowerCase.startsWith("flink-sql-client")
+    // flink.execution.target are required in Kyuubi conf currently
+    val executionTarget = conf.getOption("flink.execution.target")
+    executionTarget match {
+      case Some("yarn-application") =>
+        val buffer = new ArrayBuffer[String]()
+        buffer += flinkExecutable
+        buffer += "run-application"
+
+        // locate flink sql jars
+        val flinkExtraJars = Paths.get(flinkHome)
+          .resolve("opt")
+          .toFile
+          .listFiles(new FilenameFilter {
+            override def accept(dir: File, name: String): Boolean = {
+              name.toLowerCase.startsWith("flink-sql-client") ||
+              name.toLowerCase.startsWith("flink-sql-gateway")
+            }
+          }).map(f => f.getAbsolutePath).sorted
+
+        buffer += s"-Dpipeline.jars=${flinkExtraJars.mkString(",")}"
+        buffer += s"-Dyarn.tags=${conf.getOption(YARN_TAG_KEY).get}"
+        buffer += "-Dcontainerized.master.env.FLINK_CONF_DIR=."
```

Comment:
is it necessary?

Reply:
Yes, because flink would look for configurations in FLINK_HOME by default, but in our case we should load configurations in the container working directory.